onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -0,0 +1,622 @@
+  if (node == nullptr) {
+    return false;
+  }
+  if (node->OpType() == "Cast") {
+    const NodeAttributes& attributes = node->GetAttributes();
+    ORT_ENFORCE(attributes.find("to") != attributes.end());
+    if (attributes.at("to").i() == static_cast<int64_t> (TensorProto::FLOAT16)) {
+      std::unordered_set<NodeArg*> require_cast;
+      NodeArg* cast_input = node->MutableInputDefs()[0];
+      std::unordered_set<NodeArg*> require_type_change;
+      SearchUpstream(graph, cast_input, require_cast, require_type_change);
+      if (require_cast.size() > 0 && require_cast.find(cast_input) == require_cast.end()) {
+        // Remove Cast operation
+        VLOGS(logger, 1) << "PropagateBackwards: Removed Cast node  " << node->Name();
+        RemoveCastNodes(graph, {node});
+        InsertCastNodes(graph, require_cast, true);
+        ChangeTypeToFP16(require_type_change, logger);
+        VLOGS(logger, 1) << "PropagateBackwards: Inserted Cast nodes "
+                  << GatherNames<std::unordered_set<NodeArg*>>(require_cast);
+        VLOGS(logger, 1) << "PropagateBackwards: Changed the type from float to float16 : "
+                  << GatherNames<std::unordered_set<NodeArg*>>(require_type_change);
+        modified = true;
+      }
+    }
+  } else {
+    for (NodeArg* input: node->MutableInputDefs()) {
+      Node* producer = graph.GetMutableProducerNode(input->Name());
+      modified |= PropagateBackwards(graph, producer, logger);
+    }
+  }
+  return modified;
+}
+
+// Fuse all nodes, replace with a single node.
+// Assumptions:
+// 1. all nodes are Cast ops and are of the same Cast type
+// 2. all the nodes have the same input
+static void FuseNodes(Graph& graph, NodeArg* input, std::vector<Node*> nodes)
+{
+  std::vector<NodeArg*> outputs;
+  for (Node* node : nodes) {
+    std::vector<NodeArg*> node_outputs = node->MutableOutputDefs();
+    outputs.insert(outputs.end(), node_outputs.begin(), node_outputs.end());
+  }
+  Node* node = nodes[0];
+  (void) graph.AddNode(graph.GenerateNodeName(node->Name() + "_replace"),
+                       node->OpType(),
+                       "Created to replace a node",
+                       {input},
+                       outputs,
+                       &node->GetAttributes(),
+                       node->Domain());
+  for (Node* n : nodes) {
+    graph_utils::RemoveNodeOutputEdges(graph, *n);
+    graph.RemoveNode(n->Index());    
+  }
+}
+
+// Traverse the graph recursively searching/collecting sibling Cast op nodes to fuse and call FuseNodes.
+static bool FuseSubgraphs(Graph& graph, Node* parent, const logging::Logger& logger)
+{
+  bool modified = false;
+  for (NodeArg* output : parent->MutableOutputDefs()) {
+    std::vector<Node*> cast_fp16_siblings;
+    std::vector<Node*> cast_fp_siblings;
+    for (Node* node : graph.GetMutableConsumerNodes(output->Name())) {
+      if (node == nullptr) {
+        continue;
+      }
+      if (node->OpType() == "Cast") {
```

Comment:
I feel these two cases can be merged into one. the fused cast node will always take the child's "to" type. 
and we can use IsCastTo(Node, fp16 | fp32) here ...